### PR TITLE
Add Update and DeleteDistinct attribute message types to wasm in attribute module

### PR DESCRIPTION
**Pull request by ktalley-figure**
## Description

This PR will add Update and DeleteDistinct functionality via wasm to the attribute module.

closes: #857 
---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote unit and integration [tests](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: x/attribute/wasm/encode.go [hunk 1]
```diff
@@ -91,6 +125,34 @@ func (params *DeleteAttributeParams) Encode(contract sdk.AccAddress) ([]sdk.Msg,
...
+// The contract must be the owner of the name of the attribute being deleted.
+func (params *DeleteDistinctAttributeParams) Encode(contract sdk.AccAddress) ([]sdk.Msg, error) {
+	if err := types.ValidateAttributeAddress(params.Address); err != nil {
+		return nil, fmt.Errorf("wasm: invalid address: %w", err)
+	}
+	msg := types.NewMsgDeleteDistinctAttributeRequest(params.Address, contract, params.Name, params.Value)
+	return []sdk.Msg{msg}, nil
+}
+
+// Encode creates a MsgDeleteAttribute.
```
Comment (Taztingo) [inline x/attribute/wasm/encode.go:138]: Minor typo. I believe it should read "Encode creates a MsgUpdateAttributeRequest".

File: x/attribute/wasm/encode.go [hunk 2]
```diff
@@ -91,6 +125,34 @@ func (params *DeleteAttributeParams) Encode(contract sdk.AccAddress) ([]sdk.Msg,
 	return []sdk.Msg{msg}, nil
 }
 
+// Encode creates a MsgDeleteDistinctAttribute.
+// The contract must be the owner of the name of the attribute being deleted.
```
Comment (Taztingo) [inline x/attribute/wasm/encode.go:129]: Should we be doing this check? Is this the correct place for that check?

File: x/attribute/wasm/encode.go [hunk 3]
```diff
@@ -91,6 +125,34 @@ func (params *DeleteAttributeParams) Encode(contract sdk.AccAddress) ([]sdk.Msg,
...
+func (params *DeleteDistinctAttributeParams) Encode(contract sdk.AccAddress) ([]sdk.Msg, error) {
+	if err := types.ValidateAttributeAddress(params.Address); err != nil {
+		return nil, fmt.Errorf("wasm: invalid address: %w", err)
+	}
+	msg := types.NewMsgDeleteDistinctAttributeRequest(params.Address, contract, params.Name, params.Value)
+	return []sdk.Msg{msg}, nil
+}
+
+// Encode creates a MsgDeleteAttribute.
+// The contract must be the owner of the name of the attribute being deleted.
```
Comment (Taztingo) [inline x/attribute/wasm/encode.go:139]: Same as above.

File: x/attribute/wasm/encode.go (hunk 1 above)
Comment (ktalley-figure) [reply x/attribute/wasm/encode.go]: fixed

File: x/attribute/wasm/encode.go (hunk 2 above)
Comment (ktalley-figure) [reply x/attribute/wasm/encode.go:129]: the validation of the owner will actually happen when the message is processed. this function just creates the message to be emitted from the smart contract and only validates that the address is an actual usable address (`types.ValidateAttributeAddress`)

Comment (dwedul-figure) [reply x/attribute/wasm/encode.go:129]: The `AddAttributeParams.Encode` function has this comment too. It's not actually something specifically enforced in this wasm stuff though, it's actually enforced in the attribute module.

**Review by dwedul-figure — APPROVED** [review]

File: x/attribute/wasm/encode.go (hunk 3 above)
Comment (ktalley-figure) [reply x/attribute/wasm/encode.go]: updated comment for clarification

File: x/attribute/wasm/encode.go (hunk 2 above)
Comment (ktalley-figure) [reply x/attribute/wasm/encode.go]: updated comment for clarification

**Review by dwedul-figure — APPROVED** [review]

**Review by Taztingo — APPROVED** [review]

**Review by channa-figure — APPROVED** [review]

**Pull request merged**